### ISSUE TITLE
About v1 and v2 difference.

### ISSUE THREAD

**Issue by 1248280302** [ehmotion/eh-msc-4dof#1]
It seems that v2 increases the stroke setting. Does it mean that one section uses the torque reached signal to achieve zero point, and the other end directly defines the precise stroke? The stroke range of the v1 is that both ends use the torque reached signal to determine the stroke?

**Comment by ehmotion** [comment]
That's correct: v1 detects the stroke length whereas v2 takes it as given and considers the 'power on' position to be 0.
Position 0 is considered when the actuator is fully collapsed (as opposed to maximum position when fully extended).

Make sure to measure your actuator's stroke when using v2 and set the right value in milimeters for:
#define STROKE_LIMIT  100               //100mm limit means we do no homing

**Comment by 1248280302** [comment]
> That's correct: v1 detects the stroke length whereas v2 takes it as given and considers the 'power on' position to be 0. Position 0 is considered when the actuator is fully collapsed (as opposed to maximum position when fully extended).
> 
> Make sure to measure your actuator's stroke when using v2 and set the right value in milimeters for: #define STROKE_LIMIT 100 //100mm limit means we do no homing

Well.So v1 depends on the two torque arrival signals generated at both ends to determine the range? Does v2 need to be manually pushed down to zero before 'power on' each time?

**Comment by ehmotion** [comment]
Yes, v2 considers position 0 to be the current position when power is applied, then it 'homes' to the middle of the STROKE length, hence the need to have the actuator fully collapsed when it is powered on.
If you use it on a regular platform as vertical actuator, the platform weight takes care of the collapse of actuators but I have the feeling you don't use it as regular actuator, right? What do you use it for?

**Comment by 1248280302** [comment]
> Yes, v2 considers position 0 to be the current position when power is applied, then it 'homes' to the middle of the STROKE length, hence the need to have the actuator fully collapsed when it is powered on. If you use it on a regular platform as vertical actuator, the platform weight takes care of the collapse of actuators but I have the feeling you don't use it as regular actuator, right? What do you use it for?

Well, I want to use 4 actuators on the 3DOF simulator. Each corner of the rectangular frame is configured with one, which should be a regular shape.

**Comment by 1248280302** [comment]
@ehmotion 
I viewed the code. In addition to adding a specific travel limit, can V2 also use the torque arrival signal?
 I want to use the hall switch instead of the torque reached signal to realize automatic back to zero point after startup. Will V1 first rotate forward and reverse to detect the stroke after startup?

**Comment by ehmotion** [comment]
v1 uses the torque reached signal, v2 doesn't.
v1 extends the actuator until the signal is reached, then retracts it until the signal is reached and this is how it computes the stroke length. if you follow the arduino serial monitor output, you'll see some info about the detection process.

**Comment by 1248280302** [comment]
> > Yes, v2 considers position 0 to be the current position when power is applied, then it 'homes' to the middle of the STROKE length, hence the need to have the actuator fully collapsed when it is powered on. If you use it on a regular platform as vertical actuator, the platform weight takes care of the collapse of actuators but I have the feeling you don't use it as regular actuator, right? What do you use it for?
> 
> Well, I want to use 4 actuators on the 3DOF simulator. Each corner of the rectangular frame is configured with one, which should be a regular shape.

Well, I looked at the code and saw that there was also a safe [] interrupt in V2. I thought that V2 was to add the travel setting on the basis of V1, so I wanted to use this signal to realize automatic return to the origin when starting up.

#define SPD_HOMING 500.How is this value calculated?

#define HOMING_INC 2.What does the definition mean？

**Comment by ehmotion** [comment]
v2 is based on v1 and most of the code is still there but it is unused.
SPD_HOMING is not computed, it is empirical based on my own experience and trying to avoid breaking the actuator when detecting its limits.
you can enable limit detection with v2 if you set the STROKE_LIMIT to 0 but I have not tested this. the main use of v2 is to avoid doing limit detection and hence avoid stressing the actuator at every startup.
the safe interrupt SHOULD disable the Arduino in case an actuator limit is reached during normal use, to avoid damaging the hardware - this is not easy to test either.

**Comment by 1248280302** [comment]
@ehmotion 
In your code:
TCNT1 = period - 1;
ICR1 = period; // Timer TOP
So will a count cause an interrupt from TCNTn to ICRn?


// 3us (should not be changed, timing is critical)
const unsigned int pulseLength = 5;
Why set OCRnB=pulseLength?Does this mean that the width of a single pulse varies from 0 to 5 intervals?But why is it 3us instead of 2.5us(5 periods×0.5us)?

**Comment by ehmotion** [comment]
the original code belongs to https://github.com/cubexvr/servoController
I only adapted it for limits detection and configured it based on what is suitable for AASD15A controller.
I'm sure some answers are to be found in the ATMEGA datasheet.
I'm sorry I don't have a better answer.

**Comment by 1248280302** [comment]
@ehmotion Thanks!I'm reading the datasheet.
